### PR TITLE
[REEF-866] Refactor bridge clients to separate driver config generati…

### PR DESCRIPTION
**Pull request by anupam128**
…on from

application submission

JIRA:
  [REEF-866](https://issues.apache.org/jira/browse/REEF-866)

### CODE REVIEW

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalRuntimeDriverConfigurationGenerator.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import org.apache.reef.tang.formats.AvroConfigurationSerializer;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.io.IOException;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+final class LocalRuntimeDriverConfigurationGenerator {
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalRuntimeDriverConfigurationGenerator.java]: Please add documentation.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalRuntimeDriverConfigurationGenerator.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  private static final Logger LOG = Logger.getLogger(LocalRuntimeDriverConfigurationGenerator.class.getName());
+  private final REEFFileNames fileNames;
+  private final DriverConfigurationProvider driverConfigurationProvider;
+  private final Set<ConfigurationProvider> configurationProviders;
+  private final AvroConfigurationSerializer configurationSerializer;
+
+  @Inject
+  LocalRuntimeDriverConfigurationGenerator(final AvroConfigurationSerializer configurationSerializer,
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalRuntimeDriverConfigurationGenerator.java]: `private`?

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalRuntimeDriverConfigurationGenerator.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+                                           final DriverConfigurationProvider driverConfigurationProvider,
+                                           @Parameter(DriverConfigurationProviders.class)
+                                           final Set<ConfigurationProvider> configurationProviders){
+    this.fileNames = fileNames;
+    this.driverConfigurationProvider = driverConfigurationProvider;
+    this.configurationProviders = configurationProviders;
+    this.configurationSerializer = configurationSerializer;
+  }
+
+  public Configuration writeConfiguration(final File jobFolder,
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalRuntimeDriverConfigurationGenerator.java]: Please add documentation. Especially about the parameters.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalRuntimeDriverConfigurationGenerator.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+          .newConfigurationBuilder()
+          .bindNamedParameter(JobSubmissionDirectory.class, driverFolder.toString())
+          .build(),
+        providedConfigurations);
+    final File driverConfigurationFile = new File(driverFolder, fileNames.getDriverConfigurationPath());
+    configurationSerializer.toFile(driverConfiguration, driverConfigurationFile);
+    return driverConfiguration;
+  }
+
+  public static void main(final String[] args) throws InjectionException, IOException {
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalRuntimeDriverConfigurationGenerator.java]: Is this ever called as a program? I'd expect this to be called only from other methods?

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalRuntimeDriverConfigurationGenerator.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    final LocalSubmissionFromCS localSubmission = LocalSubmissionFromCS.fromCommandLine(args);
+    LOG.log(Level.INFO, "Local driver config generation received from C#: {0}", localSubmission);
+    final Configuration yarnConfiguration = localSubmission.getRuntimeConfiguration();
+    final LocalRuntimeDriverConfigurationGenerator localConfigurationGenerator = Tang.Factory.getTang()
+        .newInjector(yarnConfiguration)
+        .getInstance(LocalRuntimeDriverConfigurationGenerator.class);
+    localConfigurationGenerator.writeConfiguration(localSubmission.getJobFolder(),
+        localSubmission.getJobId(), ClientRemoteIdentifier.NONE);
+    System.exit(0);
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalRuntimeDriverConfigurationGenerator.java]: This shouldn't be needed here, as no threads are spawned.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalRuntimeDriverConfigurationGenerator.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        providedConfigurations);
+    final File driverConfigurationFile = new File(driverFolder, fileNames.getDriverConfigurationPath());
+    configurationSerializer.toFile(driverConfiguration, driverConfigurationFile);
+    return driverConfiguration;
+  }
+
+  public static void main(final String[] args) throws InjectionException, IOException {
+    final LocalSubmissionFromCS localSubmission = LocalSubmissionFromCS.fromCommandLine(args);
+    LOG.log(Level.INFO, "Local driver config generation received from C#: {0}", localSubmission);
+    final Configuration yarnConfiguration = localSubmission.getRuntimeConfiguration();
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalRuntimeDriverConfigurationGenerator.java]: rename to `localRuntimeConfiguration`

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnDriverConfigurationGenerator.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+ */
+final class YarnDriverConfigurationGenerator {
+  private static final Logger LOG = Logger.getLogger(YarnDriverConfigurationGenerator.class.getName());
+  private final Set<ConfigurationProvider> configurationProviders;
+  private final int driverRestartEvaluatorRecoverySeconds;
+  private final REEFFileNames fileNames;
+  private final ConfigurationSerializer configurationSerializer;
+
+  @Inject
+  YarnDriverConfigurationGenerator(@Parameter(DriverConfigurationProviders.class)
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnDriverConfigurationGenerator.java]: `private`?

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnDriverConfigurationGenerator.java [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+                                   final int driverRestartEvaluatorRecoverySeconds,
+                                   final ConfigurationSerializer configurationSerializer,
+                                   final REEFFileNames fileNames) {
+    this.configurationProviders = configurationProviders;
+    this.driverRestartEvaluatorRecoverySeconds = driverRestartEvaluatorRecoverySeconds;
+    this.fileNames = fileNames;
+    this.configurationSerializer = configurationSerializer;
+  }
+
+  public Configuration writeConfiguration(final File driverFolder,
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnDriverConfigurationGenerator.java]: Please document.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnDriverConfigurationGenerator.java [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        .set(YarnDriverConfiguration.JVM_HEAP_SLACK, 0.0)
+        .build();
+
+    final ConfigurationBuilder configurationBuilder = Tang.Factory.getTang().newConfigurationBuilder();
+    for (final ConfigurationProvider configurationProvider : this.configurationProviders) {
+      configurationBuilder.addConfiguration(configurationProvider.getConfiguration());
+    }
+    final Configuration providedConfigurations =  configurationBuilder.build();
+
+    Configuration driverConfiguration = Configurations.merge(
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnDriverConfigurationGenerator.java]: `final`?

**Comment by markusweimer** [comment]
I've done a first pass.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalRuntimeDriverConfigurationGenerator.java (hunk 4 above)
Comment (anupam128) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalRuntimeDriverConfigurationGenerator.java]: The idea is to call this from .NET for now, as intermediate step between moving this functionality out of the client completely.
In the interim we plan to handle YARN submission/communication, job resource uploading directly from .NET and just call into the config generator to generate driver config on the client.
This approach gives us the quickest way forward right now to have a proper .NET client.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnDriverConfigurationGenerator.java (hunk 9 above)
Comment (anupam128) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnDriverConfigurationGenerator.java]: we do change it.

**Comment by anupam128** [comment]
@markusweimer thanks for taking a look.

I have addressed your comments.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalRuntimeDriverConfigurationGenerator.java (hunk 4 above)
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalRuntimeDriverConfigurationGenerator.java]: Ah, OK. In that case, we should document the command line arguments and add a comment here as to where this is called from on the .NET side.That should ease future edits of this code.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnDriverConfigurationGenerator.java (hunk 9 above)
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnDriverConfigurationGenerator.java]: Kind of yes. But these configurations could also be added to the final `.merge()` call, right?

Comment (anupam128) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnDriverConfigurationGenerator.java]: The merge is in a conditional. In order to make this final I will have to make the code unnecessarily convoluted with at least one more "if". I am not sure it is worth it. Wrong conditionals are probably more error prone than mutability done with a common pattern of 
```java
foo = something
if (bar)
{
  foo = f(foo);
}

return foo;
```

My caffeinated brain might be playing tricks with me though. Can you suggest me how to do this the right way?

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalRuntimeDriverConfigurationGenerator.java (hunk 4 above)
Comment (anupam128) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalRuntimeDriverConfigurationGenerator.java]: The documentation for command line arguments was duplicated and already out of sync. I am consolidating them in one place and linking the others to that one place.

The .NET side call is going to be a subsequent PR I will add that to the documentation then.

Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/LocalRuntimeDriverConfigurationGenerator.java]: OK, Let's keep it this way for this PR, then.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnDriverConfigurationGenerator.java (hunk 9 above)
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnDriverConfigurationGenerator.java]: how about:

```java
final Configuration foo = something;
final Configuration result;
if(bar){
  result = f(foo);
}else{
  result = foo;
}

return result;
```
 Not sure whether that is cleaner to every brain, but I find it easier to parse, as `result` is only assigned one or the other way and I can easily find those ways.

But feel free to leave it as is if you find the current version easier to parse.

**Comment by markusweimer** [comment]
@anupam128 +1 on merging. However, I am not able to test right now and we'd need a volunteer for it.

**Comment by afchung** [comment]
@anupam128 @markusweimer I'll test and merge this today. Thanks!

**Pull request closed without merging**